streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala

```diff
@@ -212,6 +212,86 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
     )
   }
 
+  // helps test congestion strategies with a constant rate per Block
+  // verifies the rate of messages produced by one of the congestion strategies
+  def testBlockGeneratorCongestion(congestionStrategy: CongestionStrategy,
+                                   blockIntervalMs: Int,
+                                   maxRatePerBlock: Int,
+                                   expectedBlocks: Int,
+                                   errorBound: Double = 0.05): Unit = {
+    congestionStrategy.onBlockBoundUpdate(maxRatePerBlock)
+    val blockGeneratorListener = new FakeBlockGeneratorListenerWithStrategy(congestionStrategy)
+    val conf = new SparkConf().set("spark.streaming.blockInterval", s"${blockIntervalMs}ms")
+    val blockGenerator = new BlockGenerator(blockGeneratorListener, 1, conf)
+
+    val waitTime = expectedBlocks * blockIntervalMs
+    val expectedMessages = maxRatePerBlock * expectedBlocks
+    val expectedMessagesPerBlock = maxRatePerBlock
+    val generatedData = new ArrayBuffer[Int]
+
+    // Generate blocks
+    val startTime = System.currentTimeMillis()
+    blockGenerator.start()
+    var count = 0
+    while(System.currentTimeMillis - startTime < waitTime) {
```

Comment:
My point is that nanoTime is for tracking elapsed time. CurrentTimeMillis
is about wallclock (which can be manipulated)

-- 
Cheers,
√